### ISSUE TITLE
Add MutModule with `&mut self` in forward

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#142]
As discussed in #78, there needs to be a way for modules that need to update themselves to have `&mut self` in forward.

This would impact batch norm and also dropout.

Basically all modules would need to either:
1. implement both Module and MutModule
2. just implement MutModule

Then a model with no only MutModules (e.g. `(Linear, ReLU, Linear)`) can have forward called without forward.

However when you add a only MutModule (Dropout/BatchNorm), then your model has to be mutable to call forward.

**Comment by coreylowman** [comment]
An example implementation of this can be seen here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b95daf1c648916fe4991c4de7afbe05c

```rust
    type Model = (Affine, Dropout);
    let model: Model = (
        Affine { w: 2.0, b: 0.5 },
        Dropout {
            rng: StdRng::seed_from_u64(0),
        },
    );
    let y = model.forward(1.0);
```

results in the error:
```
error[[E0596]](https://doc.rust-lang.org/stable/error-index.html#E0596): cannot borrow `model` as mutable, as it is not declared as mutable
  --> src/main.rs:76:13
   |
70 |     let model: Model = (
   |         ----- help: consider changing this to be mutable: `mut model`
...
76 |     let y = model.forward(1.0);
   |             ^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
```

which goes away if you change model to mut. However if you keep model as non-mut and change the Dropout to a sescond Affine, it will compile fine.

**Comment by coreylowman** [comment]
Downsides of this approach are that now every module under nn has to also implement MutModule. But that should be pretty simple like this:

```rust
impl<I> MutModule<I> for <WHATEVER STRUCT>
where
    Self: Module<I>
{
    type Output = <Self as Module<I>>::Output;
    fn forward(&mut self, x: I) -> Self::Output {
        <Self as Module<I>>::forward(self, x)
    }
}
```

**Comment by coreylowman** [comment]
@vikigenius thoughts on this approach instead of just changing `Module::forward` to be `&mut`

**Comment by vikigenius** [comment]
@coreylowman can you clarify this part `Then a model with no only MutModules (e.g. (Linear, ReLU, Linear)) can have forward called without forward.`

If every module has to implement `MutModule` anyway then I am not seeing the benefit to having a separate immutable `Module` implementation

**Comment by coreylowman** [comment]
err yeah I meant forward can be called without it being mut.

Yeah after some further thought I convinced myself that that isn’t the way to go.

I’m now leaning towards adding Module::mut_forward as a method that can default to calling forward. While this may be a surprise to users coming from pytorch, I think this explicitly represents the actual behavior. pytorch’s forward hides both mutable and non mutable forward, whereas users should be able to tell when reading the code what forward will do. 

Plus library developers then know they should probably call mut_forward during training and forward for test/eval/inference.

**Comment by vikigenius** [comment]
@coreylowman Yes, I think so as well. That seems much more natural. Although I am more partial to `forward_mut` instead of `mut_forward` since everything I have seen in rust seems to have mut as a suffix rather than a prefix :)
